src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.cs

```diff
@@ -0,0 +1,357 @@
+                            Interop.Winsock.ioctlsocket(s_ipv6Socket, Interop.Winsock.IoctlSocketConstants.FIONBIO, ref blocking);
+                            s_ipv6WaitHandle = s_ipv6Socket.GetEventHandle();
+                        }
+                    }
+
+                    if ((caller != null) && (!s_callerArray.ContainsKey(caller)))
+                    {
+                        s_callerArray.Add(caller, captureContext ? ExecutionContext.Capture() : null);
+                    }
+
+                    if (s_isListening || s_callerArray.Count == 0)
+                    {
+                        return;
+                    }
+
+                    if (!s_isPending)
+                    {
+                        int length;
+                        SocketError errorCode;
+
+                        if (Socket.OSSupportsIPv4 && (startIPOptions & StartIPOptions.StartIPv4) != 0)
+                        {
+                            s_registeredWait = ThreadPool.RegisterWaitForSingleObject(
+                                s_ipv4WaitHandle,
+                                new WaitOrTimerCallback(AddressChangedCallback),
+                                StartIPOptions.StartIPv4,
+                                -1,
+                                true);
+
+                            errorCode = Interop.Winsock.WSAIoctl_Blocking(
+                                s_ipv4Socket.DangerousGetHandle(),
+                                (int)IOControlCode.AddressListChange,
+                                null, 0, null, 0,
+                                out length,
+                                SafeNativeOverlapped.Zero, IntPtr.Zero);
+
+                            if (errorCode != SocketError.Success)
+                            {
+                                NetworkInformationException exception = new NetworkInformationException();
+                                if (exception.ErrorCode != (uint)SocketError.WouldBlock)
+                                {
+                                    throw exception;
+                                }
+                            }
+
+                            SafeWaitHandle s_ipv4SocketGetEventHandleSafeWaitHandle =
+                                s_ipv4Socket.GetEventHandle().GetSafeWaitHandle();
+
+                            errorCode = Interop.Winsock.WSAEventSelect(
+                                s_ipv4Socket,
+                                s_ipv4SocketGetEventHandleSafeWaitHandle,
+                                Interop.Winsock.AsyncEventBits.FdAddressListChange);
+
+                            if (errorCode != SocketError.Success)
+                            {
+                                throw new NetworkInformationException();
+                            }
+                        }
+
+                        if (Socket.OSSupportsIPv6 && (startIPOptions & StartIPOptions.StartIPv6) != 0)
```

Comment:
Indeed, `s_registeredWait` is overwritten. There are other issues with this implementation: the `RegisteredWaitHandle` is never unregistered.

https://msdn.microsoft.com/en-us/library/w9f75h7a(v=vs.110).aspx:
> When you are finished using the RegisteredWaitHandle that is returned by this method, call its RegisteredWaitHandle.Unregister method to release references to the wait handle. We recommend that you always call the RegisteredWaitHandle.Unregister method, even if you specify true for executeOnlyOnce